drivers/syscon/syscon.c

```diff
@@ -12,106 +12,136 @@
+		break;
+	case 2:
+		*val = sys_read16(base_address + reg);
+		break;
+	case 4:
+		*val = sys_read32(base_address + reg);
+		break;
+	default:
+		return -EINVAL;
+	}
 
 	return 0;
 }
 
-int syscon_generic_write_reg(uint16_t reg, uint32_t val)
+static int syscon_generic_write_reg(const struct device *dev, uint16_t reg, uint32_t val)
 {
+	const struct syscon_generic_config *config;
 	struct syscon_generic_data *data;
 	uintptr_t base_address;
 
-	if (!syscon_dev) {
+	if (!dev) {
 		return -ENODEV;
 	}
 
-	data = syscon_dev->data;
+	data = dev->data;
+	config = dev->config;
 
-	if (sanitize_reg(&reg, data->size)) {
+	if (syscon_sanitize_reg(&reg, data->size, config->reg_width)) {
 		return -EINVAL;
 	}
 
-	base_address = DEVICE_MMIO_GET(syscon_dev);
-
-	sys_write32(val, (base_address + reg));
+	base_address = DEVICE_MMIO_GET(dev);
+
+	switch (config->reg_width) {
+	case 1:
+		sys_write8(val, (base_address + reg));
+		break;
+	case 2:
+		sys_write16(val, (base_address + reg));
+		break;
+	case 4:
+		sys_write32(val, (base_address + reg));
+		break;
+	default:
+		return -EINVAL;
+	}
 
 	return 0;
 }
 
-int syscon_generic_init(const struct device *dev)
+static int syscon_generic_get_size(const struct device *dev, size_t *size)
 {
 	struct syscon_generic_data *data = dev->data;
 
-	syscon_dev = dev;
+	*size = data->size;
+	return 0;
+}
 
-	DEVICE_MMIO_MAP(syscon_dev, K_MEM_CACHE_NONE);
+static const struct syscon_driver_api syscon_generic_driver_api = {
+	.read = syscon_generic_read_reg,
+	.write = syscon_generic_write_reg,
+	.get_base = syscon_generic_get_base,
+	.get_size = syscon_generic_get_size,
+};
 
-	data->size = DT_REG_SIZE(DT_DRV_INST(0));
+static int syscon_generic_init(const struct device *dev)
+{
+	DEVICE_MMIO_MAP(dev, K_MEM_CACHE_NONE);
 
 	return 0;
 }
 
-DEVICE_DT_INST_DEFINE(0, syscon_generic_init, NULL, &syscon_generic_data_0,
-		&syscon_generic_config_0, PRE_KERNEL_1,
-		CONFIG_SYSCON_GENERIC_INIT_PRIORITY_DEVICE, NULL);
+#define SYSCON_INIT(inst)                                                                          \
+	static const struct syscon_generic_config syscon_generic_config_##inst = {                 \
+		DEVICE_MMIO_ROM_INIT(DT_DRV_INST(inst)),                                           \
+		.reg_width = DT_INST_PROP_OR(inst, reg_io_width, 4),                               \
+	};                                                                                         \
+	static struct syscon_generic_data syscon_generic_data_##inst = {                           \
+		.size = DT_REG_SIZE(DT_DRV_INST(inst)),                                            \
+	};                                                                                         \
+	DEVICE_DT_INST_DEFINE(inst, syscon_generic_init, NULL, &syscon_generic_data_##inst,        \
+			      &syscon_generic_config_##inst, PRE_KERNEL_1,                         \
+			      CONFIG_SYSCON_INIT_PRIORITY_DEVICE, syscon_generic_driver_api);
```

Comment:
&syscon_generic_driver_api

Reply:
Thanks for the catch. I added a `build_only` test to make sure that the driver builds.

@carlocaione what are your thoughts about providing a more robust test for this? AFAICT what we would need is a section in the linker that we know the address of ahead of time. It seems that `__attribute__((at(???)))` doesn't work. Any ideas?

Reply:
you could use a `reserved-memory` node + QEMU + `syscon` to create a test. Pretty much something like https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/boards/qemu_cortex_a53/reserved_memory but adding the `syscon` on top.

Reply:
Odd, it doesn't seem to go at the address I'm specifying. I'm seeing:

```
    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/syscon/src/main.c:20: test_base_address: (addr not equal to expected_addr)
addr=0x407ff000, exoected_addr=0x42000000
```

Is that expected? I was hoping to check that `syscon_get_base()` returns the address I set via:

```
/ {
	reserved-memory {
		compatible = "reserved-memory";
		#address-cells = <1>;
		#size-cells = <1>;
		status = "okay";
		ranges;

		res0: res@42000000 {
			reg = <0x42000000 0x8>;
			label = "res0";
		};
	};
	syscon: syscon@42000000 {
		compatible = "syscon";
		status = "okay";
		reg = <0x42000000 0x8>;
		reg-io-width = <1>;
	};
};
```

Reply:
It seems that I'm also not able to write to this region. I tried:

```
uint8_t var_in_res0[DT_REG_SIZE(DT_NODELABEL(syscon))] __attribute((__section__(".res0")));

void test_main(void)
{
  printk("var_in_res0@0x%lx\n", (uintptr_t)var_in_res0);
  var_in_res0[0] = 0;
}
```

Which results in:

```
var_in_res0@0x42000000
E: ELR_ELn: 0x000000004000164c
E: ESR_ELn: 0x0000000096000046
E:   EC:  0x25 (Data Abort taken without a change in Exception level)
E:   IL:  0x1
E:   ISS: 0x46
E: FAR_ELn: 0x0000000042000000
E: TPIDRRO: 0x0100000040014990
E: x0:  0x0000000000000017  x1:  0x000000000000000a
E: x2:  0x000000004000a485  x3:  0x000000004005e599
E: x4:  0x000000000000000e  x5:  0x0000000000000002
E: x6:  0x0000000000000004  x7:  0x0000000000000000
E: x8:  0x0000000000000037  x9:  0x00000000ffffffff
E: x10: 0x0000000000000000  x11: 0x0000000000000000
E: x12: 0x0000000000000000  x13: 0x0000000000000000
E: x14: 0x0000000000000000  x15: 0x0000000000000000
E: x16: 0x0000000040005280  x17: 0x0000000000000000
E: x18: 0x0000000000000000  x30: 0x000000004000164c
E: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
E: Current thread: 0x40014640 (main)
E: Halting system
```

So this is telling me that `var_in_res` was actually placed at `0x42000000` as expected but `syscon_get_base()` is returning a different value. I also thought that this might be an artifact of `DEVICE_MMIO_IS_IN_RAM` not being set and the region being placed in ROM, but even with turning on `CONFIG_MMU=y` this is still failing.

Reply:
The address returned by `syscon_get_base()` is expected to be different from `0x42000000`, because that's a virtual address while `0x42000000` is the physical address. You are defining a reserved region specifying the physical address but the syscon driver is io-mapping that region before returning to you the virtual address, the two address can be different. You cannot directly write to `0x42000000` without triggering a fault.

What you have to do is always using the address returned by `syscon_get_base()` and nothing else. You can write a certain amount of bytes at that address with direct memory access and then you should try to get back using the sysycon driver. Or the opposite: you use the syscon driver to write some data and retrieve that back using direct memory access. Bottom line: you never access directly the memory using the physical address.

This is of course valid if you have the MMU enabled (as you should if you are using A53).

Reply:
Thanks for explaining. I've updated the test.